go/test/endtoend/vtgate/foreignkey/fk_test.go

```diff
@@ -157,3 +161,509 @@ func TestUpdateWithFK(t *testing.T) {
+				"insert into fk_t17(id, col) values (1, 7)",
+				"insert into fk_t18(id, col) values (1, 7)",
+			},
+			dmlQuery: "delete from fk_t16 where id = 1",
+			assertionQueries: []string{
+				"select * from fk_t15 order by id",
+				"select * from fk_t16 order by id",
+				"select * from fk_t17 order by id",
+				"select * from fk_t18 order by id",
+			},
+		}, {
+			name: "Delete success with cascade to delete with set null to an update set null foreign key",
+			dataQueries: []string{
+				"insert into fk_t15(id, col) values (1, 7), (2, 9)",
+				"insert into fk_t16(id, col) values (1, 7), (2, 9)",
+				"insert into fk_t17(id, col) values (1, 7)",
+				"insert into fk_t19(id, col) values (1, 7)",
+			},
+			dmlQuery: "delete from fk_t15 where id = 1",
+			assertionQueries: []string{
+				"select * from fk_t15 order by id",
+				"select * from fk_t16 order by id",
+				"select * from fk_t17 order by id",
+				"select * from fk_t19 order by id",
+			},
+		}, {
+			name: "Update success with cascade to an update set null to an update cascade foreign key",
+			dataQueries: []string{
+				"insert into fk_t15(id, col) values (1, 7), (2, 9)",
+				"insert into fk_t16(id, col) values (1, 7), (2, 9)",
+				"insert into fk_t17(id, col) values (1, 7)",
+				"insert into fk_t18(id, col) values (1, 7)",
+			},
+			dmlQuery: "update fk_t15 set col = 3 where id = 1",
+			assertionQueries: []string{
+				"select * from fk_t15 order by id",
+				"select * from fk_t16 order by id",
+				"select * from fk_t17 order by id",
+				"select * from fk_t18 order by id",
+			},
+		}, {
+			name: "Insert success for self-referenced foreign key",
+			dataQueries: []string{
+				"insert into fk_t20(id, col, col2) values (1, 7, NULL)",
+			},
+			dmlQuery: "insert into fk_t20(id, col, col2) values (2, 9, 7), (3, 10, 9)",
+			assertionQueries: []string{
+				"select * from fk_t20 order by id",
+			},
+		}, {
+			name: "Insert failure for self-referenced foreign key",
+			dataQueries: []string{
+				"insert into fk_t20(id, col, col2) values (5, 7, NULL)",
+			},
+			dmlQuery: "insert into fk_t20(id, col, col2) values (6, 9, 6)",
+			assertionQueries: []string{
+				"select * from fk_t20 order by id",
+			},
+		},
+	}
+
+	for _, tt := range testcases {
+		for _, testSharded := range []bool{false, true} {
+			t.Run(getTestName(tt.name, testSharded), func(t *testing.T) {
+				mcmp, closer := start(t)
+				defer closer()
+				// Set the correct keyspace to use from VtGates.
+				if testSharded {
+					t.Skip("Skip test since we don't have sharded foreign key support yet")
+					_ = utils.Exec(t, mcmp.VtConn, "use `ks`")
+				} else {
+					_ = utils.Exec(t, mcmp.VtConn, "use `uks`")
+				}
+
+				// Insert all the data required for running the test.
+				for _, query := range tt.dataQueries {
+					mcmp.Exec(query)
+				}
+
+				// Run the DML query that needs to be tested and verify output with MySQL.
+				_, _ = mcmp.ExecAllowAndCompareError(tt.dmlQuery)
```

Comment:
should we not check the error output here? we are comparing the errors between mysql and vitess, and if we don't want to compare them, maybe we should use a different method?

Reply:
`mcmp` has the `*testing.T` inside its struct. Inside `mcmp.ExecAllowAndCompareError` if the error from Vitess and MySQL don't match or if the query results don't match, then it fails the test anyways. 
After the verification that both MySQL and Vitess are doing the same thing (whether erroring out, or returning results which match), the function just returns the results and the error. 
As far as the test is concerned, we only want to verify we do everything exactly like MySQL and don't care whether the query fails or not. So we ignore the returned results and error.